core/src/overlay.rs

```diff
@@ -0,0 +1,426 @@
+  where
+    F: Fn(Self::Output, OverlayCloseHandle) -> O,
+    O: WidgetBuilder + 'static,
+    Self::Output: WidgetBuilder + 'static;
+
+  /// Show the widget at the give position.
+  /// if the overlay is showing, nothing will happen.
+  fn show_at(&self, pos: Point, wnd: Rc<Window>)
+  where
+    Self::Output: WidgetBuilder + 'static,
+  {
+    if self.is_show() {
+      return;
+    }
+    self.show_map(
+      move |w, _| {
+        fn_widget! {
+          @$w { anchor: Anchor::from_point(pos) }
+        }
+      },
+      wnd,
+    );
+  }
+
+  /// return whether the overlay is show.
+  fn is_show(&self) -> bool;
+
+  /// remove the showing overlay.
+  fn remove(&self);
+}
+
+pub trait OverlayBuilder {
+  type Output: WidgetBuilder + 'static;
+  fn build(&self, ctrl: OverlayCloseHandle) -> Self::Output;
+}
+
+struct OverlayData<M> {
+  builder: M,
+  style: RefCell<OverlayStyle>,
+  state: OverlayState,
+}
+
+#[derive(Clone)]
+pub struct Overlay<M>(Rc<OverlayData<M>>);
+
+impl<O, M> Overlay<M>
+where
+  M: Fn(OverlayCloseHandle) -> O + 'static,
+  O: WidgetBuilder + 'static,
+{
+  /// Create overlay from a builder with a close_handle
+  ///
+  /// ### Example
+  /// popup a widget of a button which will close when clicked.
+  /// ``` no_run
+  /// use ribir::prelude::*;
+  /// let w = fn_widget! {
+  ///   let overlay = Overlay::new(
+  ///     move |ctrl: OverlayCloseHandle| {
+  ///       let ctrl = ctrl.clone();
+  ///       fn_widget! {
+  ///         @FilledButton {
+  ///           h_align: HAlign::Center,
+  ///           v_align: VAlign::Center,
+  ///           on_tap: move |_| ctrl.remove(),
+  ///           @{ Label::new("Click to close") }
+  ///         }
+  ///       }
+  ///     },
+  ///     ctx!()
+  ///   );
+  ///   @FilledButton {
+  ///     on_tap: move |e| overlay.show(e.window()),
+  ///     @{ Label::new("Click to show overlay") }
+  ///   }
+  /// };
+  /// App::new_window(w, Some(Size::new(200., 200.)));
+  /// App::exec();
+  /// ```
+  pub fn new(builder: M, ctx: &BuildCtx) -> Self {
+    let style = RefCell::new(OverlayStyle::of(ctx));
+    Self(Rc::new(OverlayData {
+      builder,
+      style,
+      state: OverlayState::default(),
+    }))
+  }
+}
+
+impl<M> Overlay<OverlayWidget<M>>
+where
+  M: WidgetBuilder + 'static + Clone,
+{
+  /// Create overlay from Clone able widget.
+  ///
+  /// ### Example
+  ///  ``` no_run
+  ///  use ribir::prelude::*;
+  ///  let w = fn_widget! {
+  ///  let overlay = Overlay::from_widget(
+  ///     fn_widget! {
+  ///       @Text {
+  ///         h_align: HAlign::Center,
+  ///         v_align: VAlign::Center,
+  ///         text: "Hello"
+  ///       }
+  ///     },
+  ///     ctx!()
+  ///   );
+  ///   @FilledButton{
+  ///     on_tap: move |e| overlay.show(e.window()),
+  ///     @{ Label::new("Click to show overlay") }
+  ///   }
+  ///  };
+  ///  App::new_window(w, None);
+  ///  App::exec();
+  /// ```
+  pub fn from_widget(w: M, ctx: &BuildCtx) -> Self {
```

Comment:
Consider not requiring a `ctx` parameter here. Instead, query the default style in the `show` method.